### PR TITLE
New namespacing proposal

### PR DESCRIPTION
**Pull request by roji**
I played around, moving the types a bit, and this is something I came up with. This is really a basis for discussion - if you don't like any of it I have no problem throwing it away.

* Removed Logical and Physical sub-namespaces. This makes everything more discoverable - just use Npgsql.Replication and you have almost everything in Intellisense.
* Moved the pgoutput logic into NpgsqlLogicalReplicationConnection, instead of it being an extension. This is in line with pgoutput "being" logical replication, but will also simplify recycling the message instances (since they have to live somewhere).
* CreateReplicationSlot on a logical connection implicitly means pgoutput (again since it's the default). For test_decoding (and any other plugin), I'm proposing having extension methods called CreateTestDecodingReplicationSlot etc. StartReplication doesn't need this because we disambiguate via on the slot parameter type. Note that these extension methods are also in the Npgsql.Replication namespace, so that they're very discoverable.
* If you don't like this scheme, we can still have pgoutput in the names (e.g. CreatePgOutputReplicationSlot) - we can even pull it back out as extensions (just like test_decoding). But looking at the PG docs, pgoutput is simply not a term I'm seeing, and I think it would cause confusion.

### CODE REVIEW

File: src/Npgsql/Replication/NpgsqlLogicalReplicationOptions.cs [hunk 1]
```diff
@@ -1,34 +1,33 @@
...
 using System.Globalization;
 
-namespace Npgsql.Replication.Logical.Protocol
+namespace Npgsql.Replication
 {
     /// <summary>
     /// Options to be passed to the pgoutput plugin
     /// </summary>
-    public class NpgsqlPgOutputPluginOptions
+    public class NpgsqlLogicalReplicationOptions
```
Comment (Brar) [inline src/Npgsql/Replication/NpgsqlLogicalReplicationOptions.cs:10]: See comments for NpgsqlLogicalReplicationSlot

File: src/Npgsql/Replication/NpgsqlLogicalReplicationSlot.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+﻿using Npgsql.Replication.Internal;
+
+namespace Npgsql.Replication
+{
+    /// <summary>
+    /// Acts as a proxy for a logical replication slot initialized for for the logical streaming replication protocol
+    /// (pgoutput logical decoding plugin).
+    /// </summary>
+    public class NpgsqlLogicalReplicationSlot : NpgsqlLogicalReplicationSlotBase
+    {
```
Comment (Brar) [inline src/Npgsql/Replication/NpgsqlLogicalReplicationSlot.cs:10]: Here's a point where I really disagree with your naming choices.
The term "logical replication slot" is a thing in PostgreSQL and it definitely isn't tied to the "pgoutput" logical decoding plugin. I'd consider this name rather confusing than helpful because logical replication slot is the abstract base concept for all replication slots you can use with logical replication.
Since the concrete instances of replication slots are tied to the plugin they use I'd really rather name them after the plugin they use by convention. Even if the plugin name doesn't appear in the docs.
There are multiple other logical decoding plugins out there that also aren't part of the official PostgreSQL docs but if someone were to write an adapter for Npgsql I'd rather have the plugin's name appear in the adapter's class name by convention to make it at least somewhat logical and discoverable.

**Review by Brar — CHANGES_REQUESTED** [review]
In general I like the namespace simplification and I agree with your suggestions.

What I don't like is the fact that the `StartReplication()` method will now hang around in your namespace even if you don't intend to use the Logical Streaming Replication Protocol but that's cosmetics and as you wrote the message instances have to live somewhere in order to recycle them, so I find that acceptable.

What I absolutely don't like is the renaming of `NpgsqlPgOutputReplicationSlot` to `NpgsqlLogicalReplicationSlot`.
Please see my comment there.
While I agree with simplifying things and choosing defaults, I'm against obfuscating things by mixing up names from different layers of abstractions.

**Comment by roji** [comment]
> What I don't like is the fact that the StartReplication() method will now hang around in your namespace even if you don't intend to use the Logical Streaming Replication Protocol but that's cosmetics

That's true - this would apply to anyone using logical replication with a non-pgoutput plugin. Note that in this current PR, people also get StartTestEncodingReplication in their namespace. IMHO that's a good thing - any other plugin would do the same, adding an extension over NpgsqlLogicalReplicationConnection into the Npgsql.Replication namespace, and it would just light up for users the moment the nuget is referenced.

However... since you want us to rename NpgsqlLogicalReplicationSlot back to NpgsqlPgOutputReplicationSlot, we can (and probably should) also rename StartReplication to StartPgOutputReplication. This makes things more consistent, and an unused StartPgOutputReplication in your namespace is probably better than StartReplication. On the downside it's a mouthful and may be confusing to new users, but it's not too bad.

How does that sound?

> and as you wrote the message instances have to live somewhere in order to recycle them, so I find that acceptable.

It's not too hard to solve the recycling problem without moving the methods onto NpgsqlLogicalReplicationConnection (and we'd have to do it for test_encoding too, if we care enough to recycle there). The StartWhateverReplication method would simply allocate and return internal IAsyncEnumerable class, which itself would have the message instances. Then it's GetAsyncEnumerator method is the async-iterator method which contains all the logic. In other words, the question of naming and of where the methods live (on NpgsqlLogicalReplicationConnection or as extensions) are pretty much orthogonal to the question of recycling.

To summarize:

* I can rename StartReplication to StartPgOutputReplication and NpgsqlLogicalReplicationSlot to NpgsqlPgOutputReplicationSlot.
* We can also extract the pgoutput functionality back out a extension methods (just like test_encoding is currently implemented) - this seems slightly nicer for decoupling pgoutput and once again making it just another plugin.
* Recycling will work out :)

Does that sound good?

**Comment by roji** [comment]
I just pushed a commit which does what I described above - let me know what you think. It goes back closer to what you originally wanted, but uses PgOutput consistenyl everywhere (rather than Protocol) and retains the extension methods in the Npgsql.Replication namespace for easy discoverability etc.

**Comment by Brar** [comment]
Now this looks really good to me.
I really think that it's worth to accept that tiny bit of user education regarding the plugin names in order to have an API that can be applied pretty consistently across different plugins (although I agree that there probably will not be much use of the plugin model).

One thought I had recently:
What do you think about completely removing the code related to `test_decoding` from the core code base and make it a demo plugin project.
After we support `pgoutput` there's probably little reason to use it anymore and I'm a bit afraid that this code will start rotting and become a maintenance burden.
Also removing it would remove a bit of code and more importantly test-flakiness from the core project.
If we had it in a separate project we could choose do ignore or even disable it for CI.

**Review by Brar — APPROVED** [review]

**Comment by roji** [comment]
>  What do you think about completely removing the code related to test_decoding from the core code base and make it a demo plugin project.

We could...

It is already structured in the same way as an external plugin would be (extensions over LogicalReplicationConnection for creating a slot and starting replication with that slot), so from an "educational" point of view it is almost the same if it's inside or outside in a nuget.

I say let's keep it in. I don't anticipate any maintenance burden - the code is very small/simple, and even provides a bit of value via increased testing within our own repo. Another argument for keeping it is that test_decoding comes bundled with PG (right?), and our tendency is to support those out of the box with Npgsql (e.g. type handlers).

**Comment by roji** [comment]
In any case, am going to merge this and we can always split out after this PR if you want.

**Comment by roji** [comment]
Merged via 1b14b5fff20fcffa79c95cea796667667c2fbea5
